### PR TITLE
FALCON-710 , FALCON-711, FALCON-712. Process Updates to sla changes, Start Instances and end instances. Duplicate entries.

### PR DESCRIPTION
**Pull request by sandeepSamudrala**

### CODE REVIEW

File: common/src/main/java/org/apache/falcon/entity/EntityUtil.java [hunk 1]
```diff
@@ -1128,22 +1129,26 @@ public static Date getPreviousInstanceTime(Date startTime, Frequency frequency,
...
             timeZone = TimeZone.getTimeZone("UTC");
         }
-
         Date current = getPreviousInstanceTime(startTime, frequency, timeZone, startRange);
         while (true) {
+            if (!current.before(startRange) && !current.after(endRange)
+                    && current.before(endTime) && !current.before(startTime)) {
+                result.add(current);
+            }
             Date nextInstanceTime = getNextInstanceTime(current, frequency, timeZone, 1);
```
Comment (pallavi-rao) [inline common/src/main/java/org/apache/falcon/entity/EntityUtil.java]: Can assign this to current directly.

File: common/src/main/java/org/apache/falcon/entity/EntityUtil.java [hunk 2]
```diff
@@ -1128,22 +1129,26 @@ public static Date getPreviousInstanceTime(Date startTime, Frequency frequency,
...
+        if (endRange.before(startRange)) {
+            return result;
+        }
         if (timeZone == null) {
             timeZone = TimeZone.getTimeZone("UTC");
         }
-
         Date current = getPreviousInstanceTime(startTime, frequency, timeZone, startRange);
         while (true) {
+            if (!current.before(startRange) && !current.after(endRange)
```
Comment (pallavi-rao) [inline common/src/main/java/org/apache/falcon/entity/EntityUtil.java]: Shouldn't we add a check to ensure startRange and endRange is within startTime and endTime. This check will also be simplified.

Comment (sandeepSamudrala) [inline common/src/main/java/org/apache/falcon/entity/EntityUtil.java]: both startRange and endRange can be out of the start time and end time in which case the instances from the start time till endtime(Excluding end time) need to be returned)

File: common/src/main/java/org/apache/falcon/entity/EntityUtil.java (hunk 1 above)
Comment (sandeepSamudrala) [inline common/src/main/java/org/apache/falcon/entity/EntityUtil.java]: Ack. Changing it.

**Comment by sandeepSamudrala** [comment]
Merging it  as per discussion with @pallavi-rao

**Pull request merged**